File: go/blackjack/blackjack.go
package blackjack

const (
	hit   = "H"
	split = "P"
	stand = "S"
	win   = "W"
)

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	var value int
	switch card {
	case "ace":
		value = 11
	case "two":
		value = 2
	case "three":
		value = 3
	case "four":
		value = 4
	case "five":
		value = 5
	case "six":
		value = 6
	case "seven":
		value = 7
	case "eight":
		value = 8
	case "nine":
		value = 9
	case "ten", "jack", "queen", "king":
		value = 10
	default:
		value = 0
	}
	return value
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	var decision string
	var value1, value2, dealer = ParseCard(card1), ParseCard(card2), ParseCard(dealerCard)
	switch {
	case value1+value2 == 22:
		decision = split
	case value1+value2 == 21 && dealer < 10:
		decision = win
	case value1+value2 > 16:
		decision = stand
	case value1+value2 > 11 && dealer < 7:
		decision = stand
	default:
		decision = hit
	}
	return decision
}

File: ruby/blackjack/blackjack.rb
module Blackjack

  CARDS = {
    'ace' => 11,
    'two' => 2,
    'three' => 3,
    'four' => 4,
    'five' => 5,
    'six' => 6,
    'seven' => 7,
    'eight' => 8,
    'nine' => 9,
    'ten' => 10,
    'jack' => 10,
    'queen' => 10,
    'king' => 10
  }.freeze

  private_constant :CARDS

  class << self

    def parse_card(card)
      CARDS.fetch(card, 0)
    end

    def card_range(card1, card2)
      case parse_card(card1) + parse_card(card2)
      when 4..11 then 'low'
      when 12..16 then 'mid'
      when 17..20 then 'high'
      when 21 then 'blackjack'
      when 22 then 'two aces'
      end
    end

    def first_turn(card1, card2, dealer_card)
      case card_range(card1, card2)
      when 'two aces'
        'P'  # Split
      when 'low'
        'H'  # Hit
      when 'high'
        'S'  # Stand
      when 'mid'
        parse_card(dealer_card) >= 7 ? 'H' : 'S'
      when 'blackjack'
        parse_card(dealer_card) < 10 ? 'W' : 'S'
      end
    end

  end

end
